python/nav/web/devicehistory/utils/componentsearch.py

```diff
@@ -0,0 +1,129 @@
+from typing import Union
+
+from django.db.models import Model, Q, QuerySet
+from nav.models.manage import Netbox, Module, Location, NetboxGroup, Room
+
+
+def get_component_search_results(
+    search: str, button_text: str, exclude: list[type[Model]] = []
+):
+    """
+    Retrieves grouped search results for component types based on a search query.
+
+    Args:
+        search (str): The search term to filter components.
+        button_text (str): A format string for the button label, which will be formatted
+        with the component type label. It should contain exactly one '%s' placeholder.
+
+        exclude (list[Model], optional): A list of component types to exclude
+            from the search. Defaults to an empty list.
+
+    Returns:
+        dict: A dictionary where keys are component names and values are dictionaries
+            containing:
+            - 'label': The display label for the component type.
+            - 'values': The grouped or ungrouped search results.
+            - 'has_grouping': A boolean indicating if the results are grouped.
+            - 'button': The label for the associated action button.
+    """
+    if button_text.count('%s') != 1:
+        raise ValueError("button_text must contain exactly one '%s' placeholder")
+
+    results = {}
+    searches = _get_search_queries(search, exclude)
+
+    for component_type, query, group_by in searches:
+        component_query = _get_component_query(component_type, query)
+        component_results = _prefetch_and_group_components(
+            component_type, component_query, group_by
+        )
+
+        if component_results:
+            component_name = component_type._meta.db_table
+            component_label = component_type._meta.verbose_name.title()
+            results[component_name] = {
+                'label': component_label,
+                'values': component_results,
+                'has_grouping': group_by is not None,
+                'button': button_text % component_label,
+            }
+    return results
+
+
+def _get_search_queries(search: str, exclude: list[Model] = []):
+    """
+    Constructs a list of search queries for different component types.
+
+    Excludes specified component types if provided.
+    """
+    searches: list[tuple[type[Model], Q, type[Model] | None]] = [
+        (Location, Q(id__icontains=search), None),
+        (Room, Q(id__icontains=search), Location),
+        (
+            Netbox,
+            Q(sysname__icontains=search)
+            | Q(entities__device__serial__icontains=search),
+            Room,
```

Comment:
This sort-of works, but due to how vendors like Juniper screw around with their source data in `ENTITY-MIB`, a single `Netbox` may have multiple entities that claim the same serial number - so you will end up with multiple instance of the same Netbox in the result.  Some usage of `distinct()` on the resulting queryset might be in order...